File: 213.house-robber-ii.cpp
/*
 * [213] House Robber II
 *
 * https://leetcode.com/problems/house-robber-ii
 *
 * Medium (33.86%)
 * Total Accepted:
 * Total Submissions:
 * Testcase Example:  '[]'
 *
 * Note: This is an extension of House Robber.
 *
 * After robbing those houses on that street, the thief has found himself a new
 * place for his thievery so that he will not get too much attention. This
 * time, all houses at this place are arranged in a circle. That means the
 * first house is the neighbor of the last one. Meanwhile, the security system
 * for these houses remain the same as for those in the previous street.
 *
 * Given a list of non-negative integers representing the amount of money of
 * each house, determine the maximum amount of money you can rob tonight
 * without alerting the police.
 *
 * Credits:Special thanks to @Freezen for adding this problem and creating all
 * test cases.
 */
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 0) return 0;
        if(nums.size() == 1) return nums[0];
        if(nums.size() == 2) return std::max(nums[0], nums[1]);
        return std::max(rob(nums, 0, nums.size() - 2), rob(nums, 1, nums.size() - 1));
    }
    int rob(vector<int>& nums, int start, int end){
        vector<int> dp(end - start + 1, 0);
        if(end - start + 1 == 1) return nums[start];
        if(end - start + 1 == 2) return std::max(nums[start], nums[end]);
        dp[0] = nums[start];
        dp[1] = std::max(nums[start], nums[start + 1]);
        for(int i = 2; i < end - start + 1; i++){
            dp[i] = std::max(dp[i - 2] + nums[start + i], dp[i - 1]);
        }
        return dp[end - start];
    }
};

File: rust/src/leetcode/213.house-robber-ii.rs
/*
 * @lc app=leetcode id=213 lang=rust
 *
 * [213] House Robber II
 *
 * https://leetcode.com/problems/house-robber-ii/description/
 *
 * algorithms
 * Medium (38.83%)
 * Likes:    4075
 * Dislikes: 73
 * Total Accepted:    308K
 * Total Submissions: 792.5K
 * Testcase Example:  '[2,3,2]'
 *
 * You are a professional robber planning to rob houses along a street. Each
 * house has a certain amount of money stashed. All houses at this place are
 * arranged in a circle. That means the first house is the neighbor of the last
 * one. Meanwhile, adjacent houses have a security system connected, and it
 * will automatically contact the police if two adjacent houses were broken
 * into on the same night.
 *
 * Given an integer array nums representing the amount of money of each house,
 * return the maximum amount of money you can rob tonight without alerting the
 * police.
 *
 *
 * Example 1:
 *
 *
 * Input: nums = [2,3,2]
 * Output: 3
 * Explanation: You cannot rob house 1 (money = 2) and then rob house 3 (money
 * = 2), because they are adjacent houses.
 *
 *
 * Example 2:
 *
 *
 * Input: nums = [1,2,3,1]
 * Output: 4
 * Explanation: Rob house 1 (money = 1) and then rob house 3 (money = 3).
 * Total amount you can rob = 1 + 3 = 4.
 *
 *
 * Example 3:
 *
 *
 * Input: nums = [1,2,3]
 * Output: 3
 *
 *
 *
 * Constraints:
 *
 *
 * 1 <= nums.length <= 100
 * 0 <= nums[i] <= 1000
 *
 *
 */
// @lc code=start
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        match nums.len() {
            0 => 0,
            1 => nums[0],
            _ => rob_core(&nums, 0, nums.len() - 1).max(rob_core(&nums, 1, nums.len())),
        }
    }
}

fn rob_core(nums: &Vec<i32>, l: usize, r: usize) -> i32 {
    let mut rob = 0;
    let mut not_rob = 0;
    for i in l..r {
        let pre_rob = rob;
        let pre_not_rob = not_rob;
        rob = pre_not_rob + nums[i];
        not_rob = pre_rob.max(pre_not_rob);
    }
    rob.max(not_rob)
}
// @lc code=end
struct Solution;


#[test]
fn test_rob() {
    assert_eq!(Solution::rob(vec![1,2,3,1]), 4);
}
